consensus/src/traits.rs

```diff
@@ -146,10 +126,80 @@ pub trait ConsensusApi<
         })
         .await;
     }
+}
+
+/// The API that [`HotStuff`] needs to talk to the system for validating applications. This should be implemented in the `hotshot` crate and passed to all functions on `HotStuff`.
+///
+/// [`HotStuff`]: struct.HotStuff.html
+#[async_trait]
+pub trait ValidatingConsensusApi<
+    TYPES: NodeType<ConsensusType = ValidatingConsensus>,
+    LEAF: LeafType<NodeType = TYPES>,
+    I: NodeImplementation<
+        TYPES,
+        Leaf = ValidatingLeaf<TYPES>,
+        ConsensusMessage = ValidatingMessage<TYPES, I>,
+    >,
+>: ConsensusSharedApi<TYPES, LEAF, I> where
+    I::Exchanges: ValidatingExchangesType<TYPES, Message<TYPES, I>>,
+{
+    /// Send a direct message to the given recipient
+    async fn send_direct_message<PROPOSAL: ProposalType<NodeType = TYPES>, VOTE: VoteType<TYPES>>(
+        &self,
+        recipient: TYPES::SignatureKey,
+        message: ValidatingMessage<TYPES, I>,
+    ) -> std::result::Result<(), NetworkError>;
+
+    /// Send a broadcast message to the entire network.
+    async fn send_broadcast_message<
+        PROPOSAL: ProposalType<NodeType = TYPES>,
+        VOTE: VoteType<TYPES>,
+    >(
+        &self,
+        message: ValidatingMessage<TYPES, I>,
+    ) -> std::result::Result<(), NetworkError>;
+}
+
+/// The API that [`HotStuff`] needs to talk to the system for sequencing applications. This should be implemented in the `hotshot` crate and passed to all functions on `HotStuff`.
+///
+/// [`HotStuff`]: struct.HotStuff.html
```

Comment:
What does `struct.HotStuff.html` refer to?

Reply:
This was copied from an outdated comment. I'll update it!